### ISSUE TITLE
Clarify authentication possibilities and limitations for WebID-TLS and WebID-OIDC

### ISSUE THREAD

**Issue by RubenVerborgh** [solid/solid#152]

**Comment by RubenVerborgh** [comment]
(Continued from #151 where @melvincarvalho and I were discussing.)

> > Edge case indeed, but it has major architectural consequences. All current Solid client s> ftware assumes that the user is logged in with one WebID.

> What does current mean in this context?

I personally do not know any Solid software that assumes a user can be logged in to different pods with different WebIDs. Or said differently: all Solid software assumes that the user is logged in with the same WebID to all hosts.

For exampl, you can ask rdflib.js about the WebID of the current user, but this is one string, whereas reality is different for WebID-TLS.

Would you know any of our software that does think about this?

> Can you log in as different WebIDs from different origin?

With WebID-TLS as supported by current browsers, any combination of "request origin + destination origin" can have a different WebID associated with it.

For example, the user could use Web app A to access hosts B and C while using different WebIDs on B and C. The app A could detect this, but cannot change this.

In WebID-OIDC, the client can (and the current client does) ensure one WebID for access to B and C.

> In different windows (im thinking incognito windows that dont require a restart)?

No, the WebID with WebID-TLS is determined by "request origin + destination origin", window is not a factor.

> Let's nail down what's being said here.  Something like "solid auth client assumes a single WebID for a given origin, in a given browser session" -- something like this.

All Solid software apps and libraries that I know about, assume that the user is identified by one single WebID.

This assumption is correct for WebID-OIDC, but not for WebID-TLS.
In WebID-TLS in browsers, there is no concept of "the user is logged in as X". There is only a concept of "for requests from origin A to server B, the user uses identity X" for different combinations of A and B.

> WebID-OIDC was proposed to run alongside WebID-TLS, not to replace it.  So there's going to be push back if that is changed.  

I understand the pushback, but the proposal was not grounded in reality. It is not possible to run them alongside (from JavaScript client applications) as it was proposed. We need a higher level of abstraction to consider WebID-OIDC and WebID-TLS alongside in a consistent way. The old assumptions were incorrect, as issues such as https://github.com/solid/solid/issues/138#issuecomment-387439471 have revealed.

**Comment by melvincarvalho** [comment]
Thanks for the response, good progress.  First let's get an accurate statement of behavior, then look at what's needed and raise issues as necessary.

> All Solid software apps and libraries that I know about, assume that the user is identified by one single WebID.

Should this say something like, "For a given origin, at any given time, solid client side apps and libraries (that I know about) assume that the user is identified by one single WebID, independent of the destination server".  

Is that it.  And this applies to solid-auth-client, basically (I ask because we'll need to apply conclusions to others too).

So, I think that's good, the same applies to WebID-TLS as it is, and always has been used in practice.  People basically will have a default webid (good browsers let you say this is my default), and choose that for any given session.  The edge case where you might choose a different WebID for a different destination I dont think ever happens, and even if it did, is unlikely to cause any problems.  

OTOH the WebID-TLS login functionality is the most used authentication system in Solid's history.  All our code, all our apps, all our bots, for several years have relied on this functionality.  So, this is a bug.

The question is how do we enable it on the client side.  As per the previous thread, let's look at some options?

- Use another library? 
- Write inline code? 
- Write a wrapper around solid-auth-client or add a helper.
- Example : Reto wrote a nice util library that solid-auth-client retog/solid-utils#solid-utils

Ideally auth could happen in one place, so a simple helper function might be the sweet spot.  Thoughts?

**Comment by melvincarvalho** [comment]
> I understand the pushback, but the proposal was not grounded in reality. It is not possible to run them alongside (from JavaScript client applications) as it was proposed. We need a higher level of abstraction to consider WebID-TLS alongside WebID-OIDC in a consistent way. The old assumptions were incorrect, as issues such as #138 (comment) have revealed.

Thanks for writing this out, very helpful!

So, the take away is that a node solid server in OIDC mode cant currently use a node solid server in TLS mode in order to get a bearer token.  However, you dont need to use delegated authentication (which increases the attack surface), you can log in directly, say using /login/tls (or whatever the endpoint might be).  This is more secure, and could facilitate advanced use cases such as payments.  

Did I miss something?

**Comment by RubenVerborgh** [comment]
> Should this say something like, "For a given origin, at any given time, solid client side apps and libraries (that I know about) assume that the user is identified by one single WebID, independent of the destination server".

Yes.

> And this applies to solid-auth-client, basically (I ask because we'll need to apply conclusions to others too).

To all libraries, including rdflib and solid-client.
I don't know any counterexamples.
 
> OTOH the WebID-TLS login functionality is the most used authentication system in Solid's history. All our code, all our apps, all our bots, for several years have relied on this functionality. So, this is a bug.

Yes, the API of all of our client-side libraries is wrong for WebID-TLS.
I have explained this issue at length here: https://github.com/solid/solid/issues/153

This has far-reaching consequences, as I describe there. Since our client-side API for WebID-TLS is wrong, we _cannot_ realize the "alongside" effect of OIDC and TLS that we are looking for with that API. Hence why it was deemed to fail in solid-auth-client (as it did).

> The question is how do we enable it on the client side. As per the previous thread, let's look at some options?
> 
> * Use another library?
> 
> * Write inline code?
> 
> * Write a wrapper around solid-auth-client or add a helper.

As I explain in https://github.com/solid/solid/issues/153, the only perfect solutions are to change the API (into something inconvenient) or to change browser functionality through source code or extensions (hard to gain adoption).

> * Example : Reto wrote a nice util library that wraps solid-auth-client [retog/solid-utils#solid-utils](https://github.com/retog/solid-utils#solid-utils)

But that seems out of scope for this discussion: that library provides an API to work with server-side resources, but delegates authentication to solid-auth-client. I.e., solid-utils does not change anything about authentication.

> Ideally auth could happen in one place, so a simple helper function might be the sweet spot. Thoughts?

This discussion made me realize that this is impossible with the current API, which is why I wrote up https://github.com/solid/solid/issues/153.

> So, the take away is that a node solid server in OIDC mode cant currently use a node solid server in TLS mode in order to get a bearer token.

That might be confusing, because a server never uses another server for bearer tokens. Can you rephrase? Then I can comment.

> FWIW : I use WebID-TLS auth with solid on a daily basis (on my desktop machine).

Importantly, the API has _always_ been broken for WebID-TLS; we just didn't notice that for reasons I explain in https://github.com/solid/solid/issues/153.

**Comment by RubenVerborgh** [comment]
So coming back to:

> WebID-OIDC was proposed to run alongside WebID-TLS, not to replace it. So there's going to be push back if that is changed.

What this implies is that OIDC should play by the same rules as TLS, i.e., the API that worked for OIDC should also work for TLS. But we are now at a stage where OIDC and TLS are not working with the same client-side API.

And the reason for this isn't that OIDC cannot conform to this API. In fact, OIDC _does_ conform to that API. The problem is rather that TLS doesn't conform to that API, even though we thought it would. But since OIDC is facilitating usage of many pods by an application, we have started tackling such scenarios, which have revealed that, contrarily to what we thought, the API does not work for TLS.

So it's really the API we've defined for ourselves that turns out to not be compatible with TLS. OIDC is playing by the rules we set out for TLS. TLS, unfortunately, is not and cannot.
